warehouse/accounts/forms.py

```diff
@@ -117,3 +117,40 @@ def validate_password(self, field):
         if userid is not None:
             if not self.user_service.check_password(userid, field.data):
                 raise wtforms.validators.ValidationError("Invalid password.")
+
+
+class RecoverPasswordForm(LoginForm):
+
+    def __init__(self, *args, **kwargs):
+        super(ResetPasswordForm, self).__init__(*args, **kwargs)
+
+        # Instead of again defining username field, we are using LoginForm
+        # to get the username field and poping password field
+        self._fields.pop('password')
+
+
+class ResetPasswordForm(CredentialsMixin, forms.Form):
+
+    password_confirm = wtforms.PasswordField(
+        validators=[
+            wtforms.validators.DataRequired(),
+            wtforms.validators.EqualTo(
+                "password", "Passwords must match."
+            ),
+        ],
+    )
+
+    def __init__(self, *args, user_name, **kwargs):
+        super(ResetPasswordForm, self).__init__(*args, **kwargs)
+
+        self.username = user_name
+        # Instead of again defining username field, we are using LoginForm
+        # to get the username field and poping password field
+        self._fields.pop('username')
+
+    def validate_password(self, field):
+        userid = self.user_service.find_userid(self.username)
+        if userid is not None:
+            if self.user_service.check_password(userid, field.data):
+                raise wtforms.validators.ValidationError(
+                    "Password shouldn't match with previous one.")
```

Comment:
@hroncok In my view It's all about how we can limit the damage of a password leak, If the system administrator is aware of the leak then users can be notified and other steps can be taken. Since passwords are open, system administrator would expect users to change their passwords, then forbidding password reuse is quite logical: if users reuse old passwords, technically they are not really changing their passwords.

Above scenario is logical in case of password leak and an unauthorized person tried to access the system. If we think in a real user perspective,  reusing the old password means: logically he/she still remember the password. The question arises here is, Why would somebody change their password that is still there in their memory?